numpy/lib/_function_base_impl.py

```diff
@@ -1772,33 +1772,37 @@ def unwrap(p, discont=None, axis=-1, *, period=2*pi):
             540.])
     """
     p = asarray(p)
-    nd = p.ndim
     dd = diff(p, axis=axis)
-    if discont is None:
-        discont = period/2
-    slice1 = [slice(None, None)]*nd     # full slices
-    slice1[axis] = slice(1, None)
-    slice1 = tuple(slice1)
-    dtype = np.result_type(dd, period)
+    dtype = _nx.result_type(dd, period)
+    
     if _nx.issubdtype(dtype, _nx.integer):
-        interval_high, rem = divmod(period, 2)
+        half_period, rem = divmod(period, 2)
         boundary_ambiguous = rem == 0
     else:
-        interval_high = period / 2
+        half_period = period / 2
         boundary_ambiguous = True
-    interval_low = -interval_high
-    ddmod = mod(dd - interval_low, period) + interval_low
+
+    dd_mod_2period = mod(dd, 2 * period)
+    dd_periods = (dd / period).round().astype(int)
+
     if boundary_ambiguous:
-        # for `mask = (abs(dd) == period/2)`, the above line made
-        # `ddmod[mask] == -period/2`. correct these such that
-        # `ddmod[mask] == sign(dd[mask])*period/2`.
-        _nx.copyto(ddmod, interval_high,
-                   where=(ddmod == interval_low) & (dd > 0))
-    ph_correct = ddmod - dd
-    _nx.copyto(ph_correct, 0, where=abs(dd) < discont)
-    up = array(p, copy=True, dtype=dtype)
-    up[slice1] = p[slice1] + ph_correct.cumsum(axis)
-    return up
+        neg_corrections = (dd_mod_2period == half_period + period) & (dd > 0)
+        pos_corrections = (dd_mod_2period == half_period) & (dd < 0)
+        dd_periods[neg_corrections] -= 1
+        dd_periods[pos_corrections] += 1
+
+    if discont is not None and discont > half_period:
+        dd_periods[abs(dd) < discont] = 0
```

Comment:
Generally, this looks like a great start to me, but didn't wrap my head around all the changes quickly.
My main ask: We need more tests on this function (even beyond just the accuracy fix).  After that we can see if we can improve things further.

(I wonder if the cumsum being done on floats is the problem in itself.  Maybe the bigger issue may be the `ddmod - dd` line.  But going to integers or at least integral values does seem like a good idea to me!)

---

This makes me wonder if we can move this part and the "boundary handling" to earlier, via something like:
```
# May make sense to keep abs(dd) result around to not do it double.

# If the diff is equal to the period or less then discont
# we will never correct the jump, so set it to 0:
unchanged = abs(dd) == half_period
if discont is not None:
    unchanged = abs(dd) < discont

dd[unchanged] = 0
```
I may be misunderstanding the boundary corrections, but that is how I understand the intention.

---

A main thing I would like to ask for is adding a test for the fix.  Also, I checked, and I don't see any test for negative unwrapping.

Can you make sure to add test for the issue and negative values, as well as the "ambiguous boundary" cases (both float and ints?).  I know it's a bit of an ask, but changing badly tested code is not great.
(On the plus side: Tests will long outlive this code and help a lot if someone improves/rewrites it later!)

Reply:
That's a good point, I think I will be able to get to it soon. I will also add some documentation to the function that will help explaining what it promises with regards to its output and how it handles boundary cases.

I'm not sure if it will be cleaner to separate this into 2 PRs, one for testing and one this edit, or since this is only a small change there is no actual need.
Do you have thoughts on this matter?

Reply:
A separate PR can be nice to keep this more focused, but either way seems fine.  It's not like the PR is likely to become huge (or the tests particularly complex).